cpp/include/cudf/join.hpp

```diff
@@ -0,0 +1,273 @@
+ *          left_on: {0}
+ *          right_on: {1}
+ *          columns_in_common: { {0, 1} }
+ * Result: { a: {0, 1, 2, NULL}, b: {NULL, 1, 2, 3}, c: {NULL, 1, 2, 5} }
+ *
+ * @example TableA a: {0, 1, 2}
+ *          TableB b: {1, 2, 3}, c: {1, 2, 5}
+ *          left_on: {0}
+ *          right_on: {0}
+ *          columns_in_common: { }
+ * Result: { a: {0, 1, 2, NULL}, b: {NULL, 1, 2, 3}, c: {NULL, 1, 2, 5} }
+ *
+ * @throws cudf::logic_error
+ * if either table is empty or if number of rows in either table exceeds INT_MAX
+ * if number of elements in `right_on` and `left_on` are not equal
+ *
+ * @param[in] left The left table
+ * @param[in] right The right table
+ * @param[in] left_on The column's indices from `left` to join on.
+ * The column from `left` indicated by `left_on[i]` will be compared against the column 
+ * from `right` indicated by `right_on[i]`.
+ * @param[in] right_on The column's indices from `right` to join on.
+ * The column from `right` indicated by `right_on[i]` will be compared against the column 
+ * from `left` indicated by `left_on[i]`.
+ * @param[in] columns_in_common is a vector of pairs of column indices into
+ * `left_on` and `right_on`, respectively, that are "in common". For "common"
+ * columns, only a single output column will be produced, which is gathered
+ * from respective columns in `left_on` and `right_on`. Else, for every column in
+ * `left_on` and `right_on`, an output column will be produced.
+ *
+ * @returns Result of joining `left` and `right` tables on the columns
+ * specified by `left_on` and `right_on`. The resulting table will be joined columns of
+ * `left(including common columns)+right(excluding common columns)`.
+ */
+std::unique_ptr<cudf::experimental::table> full_join(
+                         cudf::table_view const& left,
+                         cudf::table_view const& right,
+                         std::vector<cudf::size_type> const& left_on,
+                         std::vector<cudf::size_type> const& right_on,
+                         std::vector<std::pair<cudf::size_type, cudf::size_type>> const& columns_in_common,
+                         cudaStream_t stream=0,
+                         rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+  /** 
+   * @brief  Performs a left semi join on the specified columns of two 
+   * tables (left, right)
+   *
+   * A left semi join only returns data from the left table, and only 
+   * returns rows that exist in the right table.
+   *
+   * @example TableA a: {0, 1, 2}
+   *          TableB b: {1, 2, 3}, a: {1, 2, 5}
+   *          left_on: {0}
+   *          right_on: {1}
+   *          return_columns: { 0 }
+   * Result: { a: {1, 2} }
+   *
+   * @example TableA a: {0, 1, 2}, c: {1, 2, 5}
+   *          TableB b: {1, 2, 3}
+   *          left_on: {0}
+   *          right_on: {0}
+   *          return_columns: { 1 }
+   * Result: { c: {1, 2} }
+   *
+   * @throws cudf::logic_error 
```

Comment:
For clarity, I think you should do one @throws line per error condition, even if the exception is the same for both.

Reply:
Done